state/watcher.go

```diff
@@ -2025,38 +2025,50 @@ func (u *Unit) WatchMeterStatus() NotifyWatcher {
 // upgrade completed field that is specific to an application name.  Used by
 // UniterAPI v9.
 func (m *Machine) WatchLXDProfileUpgradeNotifications(applicationName string) (StringsWatcher, error) {
+	if featureflag.Enabled(feature.InstanceMutater) {
+
+	}
 	unitName, err := m.LXDProfileUpgradeUnitToWatch(applicationName)
 	if err != nil {
 		return nil, errors.Trace(err)
 	}
 	watchDocId := m.instanceCharmProfileDataId(unitName)
-	filter := func(id interface{}) bool {
-		return id.(string) == watchDocId
-	}
-	returnStatus := ""
-	if featureflag.Enabled(feature.InstanceMutater) {
-		returnStatus = lxdprofile.NotRequiredStatus
-	}
-	return newInstanceCharmProfileDataWatcher(m.st, watchDocId, returnStatus, filter), nil
+	return watchInstanceCharmProfileData(m.st, watchDocId, func(value string) string {
+		return value
+	}), nil
 }
 
 // WatchLXDProfileUpgradeNotifications returns a watcher that observes the status
 // of a lxd profile upgrade by monitoring changes on the unit machine's lxd profile
 // upgrade completed field that is specific to itself.
 func (u *Unit) WatchLXDProfileUpgradeNotifications() (StringsWatcher, error) {
+	if featureflag.Enabled(feature.InstanceMutater) {
+
+	}
 	m, err := u.machine()
 	if err != nil {
 		return nil, errors.Trace(err)
 	}
 	watchDocId := m.instanceCharmProfileDataId(u.doc.Name)
+	return watchInstanceCharmProfileData(m.st, watchDocId, func(value string) string {
+		return value
+	}), nil
+}
+
+func watchInstanceCharmProfileData(backend modelBackend, watchDocId string, transform func(string) string) StringsWatcher {
+	initial := lxdprofile.NotKnownStatus
+	members := bson.D{{"_id", watchDocId}}
+	collection := instanceCharmProfileDataC
 	filter := func(id interface{}) bool {
 		return id.(string) == watchDocId
 	}
-	returnStatus := ""
-	if featureflag.Enabled(feature.InstanceMutater) {
-		returnStatus = lxdprofile.NotRequiredStatus
+	extract := func(data interface{}) (string, error) {
+		if m, ok := data.(instanceCharmProfileData); ok {
+			return m.UpgradeCharmProfileComplete, nil
+		}
+		return initial, errors.NotValidf("type")
 	}
-	return newInstanceCharmProfileDataWatcher(m.st, watchDocId, returnStatus, filter), nil
+	return newInstanceCharmProfileDataWatcher(backend, collection, members, initial, filter, extract, transform)
 }
 
 // instanceCharmProfileDataWatcher notifies about any changes to the
```

Comment:
are these comments still valid?  If the watcher has a more generic name, maybe others will be able to use it.